src/tplr/comms.py

```diff
@@ -1703,20 +1841,100 @@ async def gather(
                     f"{tplr.P(window, tplr.T() - download_start)} Downloaded peer gradients <--"
                 )
                 process_start = tplr.T()
+
+                # OPTIMIZATION: Parallelize S3 existence checks for failed downloads
+                # to avoid serial API calls that add 5s per failure
+                failed_checks = []
+                for uid, response in zip(uids, batch_responses):
+                    if isinstance(response, (Exception, type(None))):
+                        peer_bucket = self.commitments.get(int(uid))
+                        if peer_bucket:  # Only check if bucket is committed
+                            gradient_key = (
+                                f"{key}-{window}-{uid}-v{tplr.__version__}.pt"
+                            )
+                            failed_checks.append((uid, gradient_key, peer_bucket))
+
+                # Run all existence checks in parallel (huge speedup!)
+                existence_results = {}
+                if failed_checks:
+                    existence_tasks = [
+                        self.s3_object_exists(key=check[1], bucket=check[2], timeout=5)
+                        for check in failed_checks
+                    ]
+                    existence_responses = await asyncio.gather(
+                        *existence_tasks, return_exceptions=True
+                    )
+                    existence_results = {
+                        check[0]: (resp if not isinstance(resp, Exception) else False)
+                        for check, resp in zip(failed_checks, existence_responses)
+                    }
+                    tplr.logger.debug(
+                        f"Checked existence for {len(failed_checks)} failed downloads in parallel"
+                    )
+
```

Comment:
I am a bit confused, what does this help with? We already did the `get_with_retry`?